compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+  });
+
+  // Collect all contraction op users of lhs.
+  for (Operation *lhsUser : lhsUsers) {
+    if (lhsUser == seedOp.getOperation()) {
+      continue;
+    }
+
+    auto linalgUser = dyn_cast<linalg::LinalgOp>(lhsUser);
+    if (!canBeGrouped(linalgUser)) {
+      continue;
+    }
+
+    std::optional<linalg::GenericOp> userTruncOp = getTruncUser(linalgUser);
+    if (truncateOps && !userTruncOp) {
+      continue;
+    }
+
+    contractionOps.push_back(linalgUser);
+    allOps.insert(linalgUser);
+    if (truncateOps) {
+      truncateOps.value().push_back(userTruncOp.value());
+      allOps.insert(userTruncOp.value());
+    }
+  }
+
+  if (contractionOps.size() == 1) {
+    return std::nullopt;
+  }
+
+  // Check that all uses are "self-contained", i.e. one of the ops dominates
+  // all the ops (for both the contraction and the truncation ops)
+  // and that this op dominates all the other uses of all the ops.
+  std::optional<Operation *> dominatedByAllOps =
+      findDominatedByAllOps(allOps, dominanceInfo);
+  if (!dominatedByAllOps) {
+    return std::nullopt;
+  }
+
+  // Check that this operation dominates all other uses of allOps.
+  for (auto op : allOps) {
+    if (op == dominatedByAllOps.value()) {
+      continue;
+    }
+    for (Operation *user : op->getUsers()) {
+      if (allOps.contains(user)) {
+        continue;
+      }
+      if (!dominanceInfo.properlyDominates(dominatedByAllOps.value(), user)) {
+        return std::nullopt;
+      }
+    }
+  }
+  return HorizontalFusionGroup{contractionOps, truncateOps,
+                               dominatedByAllOps.value()};
+}
+
+static LogicalResult fuseGroup(RewriterBase &rewriter,
+                               HorizontalFusionGroup &fusionGroup) {
+  rewriter.setInsertionPointAfter(fusionGroup.dominatedByAll);
+
+  linalg::LinalgOp base = fusionGroup.contractionOps.front();
+  Location loc = base.getLoc();
+  auto rhsType = cast<RankedTensorType>(base->getOperand(1).getType());
+  auto outType = cast<RankedTensorType>(base->getResult(0).getType());
+  std::optional<linalg::GenericOp> baseTruncOp;
+  if (fusionGroup.truncateOps) {
+    baseTruncOp = fusionGroup.truncateOps->front();
+  }
+
+  SmallVector<ReassociationIndices> reassoc;
+  reassoc.push_back({0, 1});
+  for (int i = 0, e = rhsType.getRank() - 1; i < e; ++i) {
+    reassoc.push_back({i + 2});
+  }
+
+  SmallVector<int64_t> rhsNewShape(rhsType.getShape());
+  rhsNewShape.insert(rhsNewShape.begin(), 1);
+  RankedTensorType concatRhsType =
+      RankedTensorType::get(rhsNewShape, rhsType.getElementType());
+
+  SmallVector<Value> rhsVals;
+  for (auto op : fusionGroup.contractionOps) {
+    Value thisRhs = op.getDpsInputOperand(1)->get();
+    Value expanded = rewriter.create<tensor::ExpandShapeOp>(loc, concatRhsType,
+                                                            thisRhs, reassoc);
+    rhsVals.push_back(expanded);
+  }
+
+  Value newRhs = rewriter.create<tensor::ConcatOp>(loc, 0, rhsVals);
```

Comment:
style nit: add comments for `0`.

```suggestion
  Value newRhs = rewriter.create<tensor::ConcatOp>(loc, /*dim=*/0, rhsVals);
```